sdlib/d/gc/slab.d

```diff
@@ -389,44 +402,64 @@ unittest SlabAllocInfo {
 }
 
 /**
- * Packed Free Space is stored as a 14-bit unsigned integer, in one or two bytes:
+ * Packed Free Space is stored as a 14-bit unsigned integer, with 2 bits for flags.
+ *
+ * 15     8 7      0
+ * sfvvvvvv vvvvvvvv
+ *
+ * s: small bit set if length == 1
+ * f: finalizer bit = 1 if a finalizer is stored in allocation
+ * v: free space, only set if length > 1
+ *
+ * If free space is only 1 byte, then the lower byte of the 16-bit value is
+ * used by the allocation itself, and is not used while reading. Nor is it set
+ * when writing the upper byte
  *
- * /---- byte at ptr ----\ /-- byte at ptr + 1 --\
- * B7 B6 B5 B4 B3 B2 B1 B0 A7 A6 A5 A4 A3 A2 A1 A0
- * \_______14 bits unsigned integer________/  \  \_ Set if and only if B0..B7 used.
- *                                             \_ Set when finalizer is present;
- *                                                preserved when writing.
+ * Note: little endian only! The lower 8 bits are stored first in memory.
+ * For big endian support, we likely have to change the end where the bits go.
  */
 static assert(MaxSmallSize < 0x4000,
               "Max small alloc size doesn't fit in 14 bits!");
 
+enum FinalizerBit = 1 << 14;
+enum SingleByteBit = 1 << 15;
+enum FreeSpaceMask = ushort.max & ~(FinalizerBit | SingleByteBit);
+
 ushort readPackedFreeSpace(ushort* ptr) {
```

Comment:
It might be worth returning a uint here. It would avoid the need for a mask in the end.

Reply:
Makes sense actually to return a `size_t`. Since `writePackedFreeSpace` accepts one.